### PR TITLE
As a data engineer, I want a crawler EU covid response timeline

### PR DESCRIPTION
**Pull request by cosletmihai**

### CODE REVIEW

File: eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                month_name = block.xpath('*//h2/text()').get()
+
+            else:
+                month_timeline = block.xpath('*//li')
+                for month in month_timeline:
+                    date = month.xpath('*[@class="timeline__list__item__title"]/text()').get()
+                    title = month.xpath('*//h4/text() | *//h4/strong/text()').get()
+                    body = ' '.join(month.xpath('*//p[string-length(text()) > 0]').extract())
+
+                    presscorner_base_url = 'https://ec.europa.eu/commission/presscorner/detail'
```
Comment (costezki) [inline eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py:35]: better to make it a constant somewhere visible

File: eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                    presscorner_links = [link.attrib['href'] for link in month.xpath('*//p//a') if
```
Comment (costezki) [inline eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py:36]: I would make a dedicated function for checking this. Although you may argue that this is an overkill.

File: eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                        meta = {}
+                        meta['month_name'] = month_name
+                        meta['date'] = date
+                        meta['title'] = title
+                        meta['excerpt'] = body
+                        meta['presscorner_links'] = presscorner_links
+                        meta['all_links'] = [link.attrib['href'] for link in month.xpath('*//p//a')]
+                        yield SplashRequest(url=presscorner_link, callback=self.parse_presscorner_page,
+                                            args={'wait': 5},
+                                            meta=meta)
```
Comment (costezki) [inline eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py:50]: so all the attributes parsed in the parent are passed to the child?! 
How interesting.

File: eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    def parse_presscorner_page(self, response):
+        meta = response.meta
+        item = EuActionTimelineItem()
```
Comment (costezki) [inline eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py:54]: you may want to define an item initialiser with `**kwargs`; and then just unpack the meta there
`EuActionTimelineItem(**meta)`

File: eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        item['presscorner_links'] = meta['presscorner_links']
+        item['all_links'] = meta['all_links']
+        item['article_content'] = response.xpath('//div[@class="ecl-paragraph"]').get()
+        item['detail_title'] = response.xpath(
+            '//h1[@class="ecl-heading ecl-heading--h1 ecl-u-color-white"]//text()').extract()
+
+        detail_links_start = response.xpath('//p[contains(., "For More Information")]')
+        if detail_links_start:
+            item['for_more_information_links'] = [link.attrib['href'] for link in
+                                                  detail_links_start[0].xpath('following-sibling::p/a')]
```
Comment (costezki) [inline eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py:68]: wow, this is a real item fishing. But seems to be the only option.

File: eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        item['date'] = meta['date']
+        item['title'] = meta['title']
+        item['excerpt'] = meta['excerpt']
+        item['presscorner_links'] = meta['presscorner_links']
+        item['all_links'] = meta['all_links']
+        item['article_content'] = response.xpath('//div[@class="ecl-paragraph"]').get()
+        item['detail_title'] = response.xpath(
+            '//h1[@class="ecl-heading ecl-heading--h1 ecl-u-color-white"]//text()').extract()
+
+        detail_links_start = response.xpath('//p[contains(., "For More Information")]')
```
Comment (costezki) [inline eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py:65]: Wow! certainly predictable, but unfortunately nothing more reliable is available.

**Review by costezki — APPROVED** [review]
well done! 
I like the outcome.

File: Makefile [hunk 7]
```diff
@@ -9,10 +9,20 @@ install:
 	@ pip install --upgrade pip
 	@ pip install -r resources/requirements.txt
 
+install-scrapy-dependencies:
```
[the remark this answers is not in the document]
Comment (cosletmihai) [reply Makefile:12]: I was thinking that it's better to have separate "concerns". doesn't really make sense to always install scrapy if you're gonna use another part of the project

File: eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+    def start_requests(self):
+        yield scrapy.Request(url=self.url, callback=self.parse_main_page, headers=self.headers)
+
+    def parse_main_page(self, response):
+        data = {}
+        timeline_data = response.xpath(
+            '//div[@class="field field-name-field-core-timelines field--field-core-timelines"]/div["field__items"]/*[ @class !="clearfix"]')
+        month_name = ''
+        for index, block in enumerate(timeline_data):
+            if not index % 2:
```
[the remark this answers is not in the document]
Comment (cosletmihai) [reply eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py:24]: indeed

File: eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                self.logger.info(f'Processing data for {month_name}.')
+
+                month_name = block.xpath('*//h2/text()').get()
```
[the remark this answers is not in the document]
Comment (cosletmihai) [reply eu_action_timeline/eu_action_timeline/spiders/eu_timeline_spider.py:27]: it means to search only for descendants. If I don't use `*`, it gives me results from the whole page

**Pull request merged**